### PR TITLE
rustdoc: Remove OpaqueTy

### PR DESCRIPTION
**Pull request by aDotInTheVoid**
r? @ghost

Apparently this works lol?!?

try-job: aarch64-apple

### CODE REVIEW

**Review by GuillaumeGomez — APPROVED** [review]
I love when code can just be removed. :)

**Comment by fmease** [comment]
Good catch! Youp, it's fine since we don't care about the internal opaque *items*, we just care about the opaque *types*.

File: src/rustdoc-json-types/lib.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@ use serde::{Deserialize, Serialize};
 use std::path::PathBuf;
 
 /// rustdoc format-version.
-pub const FORMAT_VERSION: u32 = 30;
+pub const FORMAT_VERSION: u32 = 31;
```
Comment (aDotInTheVoid) [inline src/rustdoc-json-types/lib.rs:11]: This shouldn't me merged until after #127289, as they both bump the version

File: src/librustdoc/formats/item_type.rs [hunk 2]
```diff
@@ -52,7 +52,7 @@ pub(crate) enum ItemType {
     AssocConst = 19,
     Union = 20,
     ForeignType = 21,
-    OpaqueTy = 22,
+    // OpaqueTy used to be here, but it was removed in #127276
```
Comment (aDotInTheVoid) [inline src/librustdoc/formats/item_type.rs:55]: CC @notriddle, as this is part of the rustdoc-the-binary<->rustdoc-search boundry

**Comment by camelid** [comment]
Awesome find, thanks!

@bors r+ rollup

**Comment by jieyouxu** [comment]
This PR looks suspicious to me as the cause of failure in https://github.com/rust-lang/rust/pull/127475#issuecomment-2213006973. Feel free to re-approve if this PR is not the cause.

@bors r-

**Comment by jieyouxu** [comment]
<details>
<summary>ICE message</summary>

```
2024-07-08T04:32:52.1458770Z thread 'rustc' panicked at src/librustdoc/clean/mod.rs:2819:18:
2024-07-08T04:32:52.1560090Z internal error: entered unreachable code: not yet converted
2024-07-08T04:32:52.1663130Z stack backtrace:
2024-07-08T04:32:52.1742000Z    0:        0x10603eac0 - <std::sys::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h03c274328d9923fb
2024-07-08T04:32:52.1745950Z    1:        0x106081d54 - core::fmt::write::hfc23d7de2930d59b
2024-07-08T04:32:52.1809130Z    2:        0x1060349fc - std::io::Write::write_fmt::h8ff76bf83f9ff955
2024-07-08T04:32:52.1872480Z    3:        0x10603e918 - std::sys::backtrace::print::hb7f4775d7ed1e12e
2024-07-08T04:32:52.1892800Z    4:        0x106040f28 - std::panicking::default_hook::{{closure}}::h33f7b74c09dbaa9d
2024-07-08T04:32:52.1899670Z    5:        0x106040bf4 - std::panicking::default_hook::hdd08ba5aa598c81b
2024-07-08T04:32:52.1904960Z    6:        0x10f1b0690 - <alloc[1e9c988f37cf81b2]::boxed::Box<rustc_driver_impl[a7993e430336e74c]::install_ice_hook::{closure#0}> as core[cc103daf103e831a]::ops::function::Fn<(&dyn for<'a, 'b> core[cc103daf103e831a]::ops::function::Fn<(&'a std[b7d054e4fd31b525]::panic::PanicHookInfo<'b>,), Output = ()> + core[cc103daf103e831a]::marker::Sync + core[cc103daf103e831a]::marker::Send, &std[b7d054e4fd31b525]::panic::PanicHookInfo)>>::call
2024-07-08T04:32:52.1979680Z    7:        0x106041ab8 - std::panicking::rust_panic_with_hook::hcdca9952db3e7a6c
2024-07-08T04:32:52.1980790Z    8:        0x1060413a0 - std::panicking::begin_panic_handler::{{closure}}::h7530693faefe330d
2024-07-08T04:32:52.1981750Z    9:        0x10603ef4c - std::sys::backtrace::__rust_end_short_backtrace::h5d8a6c5af936bc6d
2024-07-08T04:32:52.1996180Z   10:        0x106041090 - _rust_begin_unwind
2024-07-08T04:32:52.2054110Z   11:        0x10609c66c - core::panicking::panic_fmt::hf4041e2666da7a59
2024-07-08T04:32:52.2055530Z   12:        0x10467a434 - <rustdoc[6e5addee39934cb1]::core::DocContext>::with_param_env::<alloc[1e9c988f37cf81b2]::vec::Vec<rustdoc[6e5addee39934cb1]::clean::types::Item>, rustdoc[6e5addee39934cb1]::clean::clean_maybe_renamed_item::{closure#1}>
2024-07-08T04:32:52.2057830Z   13:        0x10482a88c - <&mut rustdoc[6e5addee39934cb1]::clean::clean_doc_module::{closure#2} as core[cc103daf103e831a]::ops::function::FnOnce<(&(&rustc_hir[faf7167ecb130f2c]::hir::Item, core[cc103daf103e831a]::option::Option<rustc_span[fb1ade04d581d706]::symbol::Symbol>, core[cc103daf103e831a]::option::Option<rustc_span[fb1ade04d581d706]::def_id::LocalDefId>),)>>::call_once
2024-07-08T04:32:52.2062520Z   14:        0x1046d82cc - <alloc[1e9c988f37cf81b2]::vec::Vec<rustdoc[6e5addee39934cb1]::clean::types::Item> as alloc[1e9c988f37cf81b2]::vec::spec_extend::SpecExtend<rustdoc[6e5addee39934cb1]::clean::types::Item, core[cc103daf103e831a]::iter::adapters::flatten::FlatMap<indexmap[e87678ff1e8617bb]::map::iter::Values<(rustc_span[fb1ade04d581d706]::def_id::LocalDefId, core[cc103daf103e831a]::option::Option<rustc_span[fb1ade04d581d706]::symbol::Symbol>), (&rustc_hir[faf7167ecb130f2c]::hir::Item, core[cc103daf103e831a]::option::Option<rustc_span[fb1ade04d581d706]::symbol::Symbol>, core[cc103daf103e831a]::option::Option<rustc_span[fb1ade04d581d706]::def_id::LocalDefId>)>, alloc[1e9c988f37cf81b2]::vec::Vec<rustdoc[6e5addee39934cb1]::clean::types::Item>, rustdoc[6e5addee39934cb1]::clean::clean_doc_module::{closure#2}>>>::spec_extend
2024-07-08T04:32:52.2065860Z   15:        0x10483ba18 - rustdoc[6e5addee39934cb1]::clean::clean_doc_module
2024-07-08T04:32:52.2067110Z   16:        0x10482931c - <&mut rustdoc[6e5addee39934cb1]::clean::clean_doc_module::{closure#1} as core[cc103daf103e831a]::ops::function::FnMut<(&rustdoc[6e5addee39934cb1]::visit_ast::Module,)>>::call_mut
2024-07-08T04:32:52.2069470Z   17:        0x1046d7990 - <alloc[1e9c988f37cf81b2]::vec::Vec<rustdoc[6e5addee39934cb1]::clean::types::Item> as alloc[1e9c988f37cf81b2]::vec::spec_extend::SpecExtend<rustdoc[6e5addee39934cb1]::clean::types::Item, core[cc103daf103e831a]::iter::adapters::filter_map::FilterMap<core[cc103daf103e831a]::slice::iter::Iter<rustdoc[6e5addee39934cb1]::visit_ast::Module>, rustdoc[6e5addee39934cb1]::clean::clean_doc_module::{closure#1}>>>::spec_extend
2024-07-08T04:32:52.2071300Z   18:        0x10483b9f0 - rustdoc[6e5addee39934cb1]::clean::clean_doc_module
2024-07-08T04:32:52.2072420Z   19:        0x10482931c - <&mut rustdoc[6e5addee39934cb1]::clean::clean_doc_module::{closure#1} as core[cc103daf103e831a]::ops::function::FnMut<(&rustdoc[6e5addee39934cb1]::visit_ast::Module,)>>::call_mut
2024-07-08T04:32:52.2074910Z   20:        0x1046d7990 - <alloc[1e9c988f37cf81b2]::vec::Vec<rustdoc[6e5addee39934cb1]::clean::types::Item> as alloc[1e9c988f37cf81b2]::vec::spec_extend::SpecExtend<rustdoc[6e5addee39934cb1]::clean::types::Item, core[cc103daf103e831a]::iter::adapters::filter_map::FilterMap<core[cc103daf103e831a]::slice::iter::Iter<rustdoc[6e5addee39934cb1]::visit_ast::Module>, rustdoc[6e5addee39934cb1]::clean::clean_doc_module::{closure#1}>>>::spec_extend
2024-07-08T04:32:52.2076760Z   21:        0x10483b9f0 - rustdoc[6e5addee39934cb1]::clean::clean_doc_module
2024-07-08T04:32:52.2077460Z   22:        0x10495f294 - rustdoc[6e5addee39934cb1]::clean::utils::krate
2024-07-08T04:32:52.2078540Z   23:        0x10480d0d4 - <rustc_session[35e24ab0d5bb60b]::session::Session>::time::<rustdoc[6e5addee39934cb1]::clean::types::Crate, rustdoc[6e5addee39934cb1]::core::run_global_ctxt::{closure#4}>
2024-07-08T04:32:52.2079600Z   24:        0x10467d900 - rustdoc[6e5addee39934cb1]::core::run_global_ctxt
2024-07-08T04:32:52.2080880Z   25:        0x10480f72c - <rustc_middle[71274973e736d054]::ty::context::GlobalCtxt>::enter::<rustdoc[6e5addee39934cb1]::main_args::{closure#1}::{closure#0}::{closure#0}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>
2024-07-08T04:32:52.2082720Z   26:        0x1047fdec8 - rustc_interface[8e82f8070106c52]::interface::run_compiler::<core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>, rustdoc[6e5addee39934cb1]::main_args::{closure#1}>::{closure#1}
2024-07-08T04:32:52.2086490Z   27:        0x1047c4b38 - <scoped_tls[5925241c2ed4484d]::ScopedKey<rustc_span[fb1ade04d581d706]::SessionGlobals>>::set::<rustc_interface[8e82f8070106c52]::util::run_in_thread_with_globals<rustc_interface[8e82f8070106c52]::util::run_in_thread_pool_with_globals<rustc_interface[8e82f8070106c52]::interface::run_compiler<core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>, rustdoc[6e5addee39934cb1]::main_args::{closure#1}>::{closure#1}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#0}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>
2024-07-08T04:32:52.2091970Z   28:        0x104898244 - rustc_span[fb1ade04d581d706]::create_session_globals_then::<core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>, rustc_interface[8e82f8070106c52]::util::run_in_thread_with_globals<rustc_interface[8e82f8070106c52]::util::run_in_thread_pool_with_globals<rustc_interface[8e82f8070106c52]::interface::run_compiler<core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>, rustdoc[6e5addee39934cb1]::main_args::{closure#1}>::{closure#1}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#0}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
2024-07-08T04:32:52.2097230Z   29:        0x1047f7c58 - std[b7d054e4fd31b525]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[8e82f8070106c52]::util::run_in_thread_with_globals<rustc_interface[8e82f8070106c52]::util::run_in_thread_pool_with_globals<rustc_interface[8e82f8070106c52]::interface::run_compiler<core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>, rustdoc[6e5addee39934cb1]::main_args::{closure#1}>::{closure#1}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#0}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>
2024-07-08T04:32:52.2102910Z   30:        0x10466f0e8 - <<std[b7d054e4fd31b525]::thread::Builder>::spawn_unchecked_<rustc_interface[8e82f8070106c52]::util::run_in_thread_with_globals<rustc_interface[8e82f8070106c52]::util::run_in_thread_pool_with_globals<rustc_interface[8e82f8070106c52]::interface::run_compiler<core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>, rustdoc[6e5addee39934cb1]::main_args::{closure#1}>::{closure#1}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#0}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[cc103daf103e831a]::result::Result<(), rustc_span[fb1ade04d581d706]::ErrorGuaranteed>>::{closure#2} as core[cc103daf103e831a]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
2024-07-08T04:32:52.2106870Z   31:        0x10604a004 - std::sys::pal::unix::thread::Thread::new::thread_start::habdd2bc86a98b64c
2024-07-08T04:32:52.2107520Z   32:        0x1854bef94 - __pthread_joiner_wake
2024-07-08T04:32:52.2107920Z 
2024-07-08T04:32:52.2109420Z error: the compiler unexpectedly panicked. this is a bug.
2024-07-08T04:32:52.2117200Z 
2024-07-08T04:32:52.2118360Z note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-rustdoc&template=ice.md
2024-07-08T04:32:52.2119160Z 
2024-07-08T04:32:52.2119550Z note: please make sure that you have updated to the latest nightly
2024-07-08T04:32:52.2119990Z 
2024-07-08T04:32:52.2120650Z note: please attach the file at `/Users/runner/work/rust/rust/rustc-ice-2024-07-08T04_32_52-20708.txt` to your bug report
2024-07-08T04:32:52.2121340Z 
2024-07-08T04:32:52.2122550Z note: compiler flags: --crate-type lib -Z unstable-options -C symbol-mangling-version=v0 -Z unstable-options -Z unstable-options -Z normalize-docs -Z crate-attr=warn(rust_2018_idioms) -Z force-unstable-if-unmarked -Z unstable-options
2024-07-08T04:32:52.2123700Z 
2024-07-08T04:32:52.2124010Z note: some of the compiler flags provided by cargo are hidden
2024-07-08T04:32:52.2124480Z 
2024-07-08T04:32:52.2124670Z query stack during panic:
2024-07-08T04:32:52.2125090Z end of query stack
2024-07-08T04:32:52.2125940Z [1m[31merror[0m[1m:[0m could not document `rustc_mir_transform`
2024-07-08T04:32:52.2126700Z [1m[33mwarning[0m[1m:[0m build failed, waiting for other jobs to finish...
```
</details>

File: src/librustdoc/clean/mod.rs [hunk 3]
```diff
@@ -2735,10 +2735,6 @@ fn clean_maybe_renamed_item<'tcx>(
                 Box::new(clean_ty(ty, cx)),
                 Constant { kind: ConstantKind::Local { body: body_id, def_id } },
             ),
-            ItemKind::OpaqueTy(ref ty) => OpaqueTyItem(OpaqueTy {
```
Comment (camelid) [inline src/librustdoc/clean/mod.rs:2738]: Seems like we need to explicitly handle this and return an empty vec instead.

Comment (camelid) [reply src/librustdoc/clean/mod.rs:2738]: A bit strange that our tests didn't catch this. Do you mind seeing if you can add a test that would have discovered this ICE?

File: src/librustdoc/clean/mod.rs [hunk 4]
```diff
@@ -2735,10 +2735,8 @@ fn clean_maybe_renamed_item<'tcx>(
...
-                bounds: ty.bounds.iter().filter_map(|x| clean_generic_bound(x, cx)).collect(),
-                generics: clean_generics(ty.generics, cx),
-            }),
+            // Not sure exactly why this is ok??
```
Comment (camelid) [inline src/librustdoc/clean/mod.rs:2738]: So I think basically this function transforms a hir item into zero or more clean items. If we want to ignore the hir item, we just return an empty list to "delete" it. Unless there was something else you were asking about?

Comment (aDotInTheVoid) [reply src/librustdoc/clean/mod.rs:2755]: > Unless there was something else you were asking about?

Yeah, I'm more wondering why removing the item was fine. But I think notriddles answer above classifies that for me

**Review by camelid — APPROVED** [review]
Thanks again!

**Comment by camelid** [comment]
r=me with conflicts fixed

**Pull request merged**